File: agenthub/callbacks.py
"""Frontend callback protocol for Agent Hub.

Defines the interface that a frontend (Discord, CLI, etc.) implements
to receive notifications from the agent orchestration layer.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol


class PostMessageFn(Protocol):
    async def __call__(self, agent_name: str, text: str) -> None: ...


class PostSystemFn(Protocol):
    async def __call__(self, agent_name: str, text: str) -> None: ...


class OnLifecycleFn(Protocol):
    async def __call__(self, agent_name: str) -> None: ...


class OnSessionIdFn(Protocol):
    async def __call__(self, agent_name: str, session_id: str) -> None: ...


class GetChannelFn(Protocol):
    async def __call__(self, agent_name: str) -> Any: ...


@dataclass
class FrontendCallbacks:
    """Callbacks from Agent Hub to the frontend layer.

    All callbacks are async. The hub calls these to notify the frontend
    about agent lifecycle events and to send messages to users.
    """

    post_message: PostMessageFn  # Send a message to the agent's channel
    post_system: PostSystemFn  # Send a system notification
    on_wake: OnLifecycleFn  # Agent woke up
    on_sleep: OnLifecycleFn  # Agent went to sleep
    on_session_id: OnSessionIdFn  # Persist session ID
    get_channel: GetChannelFn  # Get frontend channel for an agent

File: axi-rs/crates/axi-hub/src/callbacks.rs
//! Frontend callback trait for Agent Hub.
//!
//! Defines the interface that a frontend (Discord, CLI, etc.) implements
//! to receive notifications from the agent orchestration layer.

use std::future::Future;
use std::pin::Pin;

/// Result type for async callbacks.
pub type CallbackResult = Pin<Box<dyn Future<Output = ()> + Send>>;

/// Frontend callbacks — the hub calls these to notify the frontend
/// about agent lifecycle events and to send messages to users.
///
/// All methods are async. We use trait objects rather than generics
/// so the hub can store a single Box<dyn FrontendCallbacks>.
pub trait FrontendCallbacks: Send + Sync {
    // Core messaging
    fn post_message(&self, agent_name: &str, text: &str) -> CallbackResult;
    fn post_system(&self, agent_name: &str, text: &str) -> CallbackResult;

    // Lifecycle events
    fn on_wake(&self, agent_name: &str) -> CallbackResult;
    fn on_sleep(&self, agent_name: &str) -> CallbackResult;
    fn on_session_id(&self, agent_name: &str, session_id: &str) -> CallbackResult;

    // Session events
    fn on_spawn(&self, agent_name: &str) -> CallbackResult;
    fn on_kill(&self, agent_name: &str, session_id: Option<&str>) -> CallbackResult;

    // Broadcast
    fn broadcast(&self, message: &str) -> CallbackResult;
    fn schedule_rate_limit_expiry(&self, delay_secs: f64);

    // Idle monitoring
    fn on_idle_reminder(&self, agent_name: &str, idle_minutes: f64) -> CallbackResult;

    // Hot restart
    fn on_reconnect(&self, agent_name: &str, was_mid_task: bool) -> CallbackResult;

    // Shutdown
    fn close_app(&self) -> CallbackResult;
    fn kill_process(&self) -> CallbackResult;
    fn send_goodbye(&self) -> CallbackResult;
}
